plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -347,12 +347,50 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
 	return selector.PodSelectorMatches
 }
 
+// New versions of the scheduler will apply their scheduling predicates
+// to both hardNodeAffinity and nodeSelector,
+// i.e. the pod can only schedule onto nodes that are in both sets.
+// We will not attempt to convert between hardNodeAffinity and nodeSelector.
 func PodMatchesNodeLabels(pod *api.Pod, node *api.Node) bool {
-	if len(pod.Spec.NodeSelector) == 0 {
+	if len(pod.Spec.NodeSelector) == 0 && (pod.Spec.Affinity == nil || pod.Spec.Affinity.HardNodeAffinity == nil) {
 		return true
 	}
-	selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
-	return selector.Matches(labels.Set(node.Labels))
+
+	// Set nodeSelectorMatches default to true to avoid changing the final match result,
+	// will be updated to exact match result if nodeSelector is specified.
+	nodeSelectorMatches := true
+	if len(pod.Spec.NodeSelector) > 0 {
+		selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
+		nodeSelectorMatches = selector.Matches(labels.Set(node.Labels))
+	}
+
```

Comment:
You could do
```
if (!nodeSelectorMatches)
     return false
```
here since you know the function can never return true in that case.